### ISSUE TITLE
Issue with "drop the two-handed"

### ISSUE THREAD

**Issue by WibbenZ** [otland/forgottenserver#2101]
### Before creating an issue, please ensure:
- [x] This is a bug in the software that resides in this repository, and not a
      support matter (use https://otland.net/forums/support.16/ for support)
- [x] This issue is reproducible without changes to the code in this repository

### Steps to reproduce (include any configuration/script required to reproduce)
1. Equip a two handed weapon
2. Equip a one handed weapon or shield

### Expected behaviour

The two handed weapon should be moved to your backpack, while the shield / other weapon should be equiped.
Same if you got a shield and a one handed weapon it should move both to your backpack and equip the two handed weapon.

### Actual behaviour

You get a message telling you to move the two handed weapon

### Environment

TFS 1.3

**Comment by ranisalt** [comment]
Isn't there a PR here handling this already?

**Comment by WibbenZ** [comment]
Oh gg didn't notice, he never referenced it to the issue :P 
#1688

**Comment by wgrriffel** [comment]
i don't think #1688 is a proper solution, it's a work around, datapack are suppose to edit each server personal stuff, not fix game problems, imo

**Comment by WibbenZ** [comment]
Well this server is based of a game and that game has that feature.

**Comment by ranisalt** [comment]
I agree with @wgrriffel, as I previously stated a clumsy developer can break the server only messing with a script while it's not so easy and requires much more knowledge to break the same thing messing with the engine.

**Comment by WibbenZ** [comment]
Okay well then we have to re do it, since that guy seems to be inactive.

**Comment by jo3bingham** [comment]
I may be wrong, but the message you're getting should be coming from here: https://github.com/otland/forgottenserver/blob/master/src/game.cpp#L1102

Which means the return message is being generated from the queryAdd() call above: https://github.com/otland/forgottenserver/blob/master/src/game.cpp#L1068

Tibia works by only letting you equip a two-handed weapon if it's inside a container that you own. This doesn't include containers that you cannot manually place items inside of (e.g., Inbox, Store Inbox, etc.) You cannot equip a two-handed weapon from the ground, or a container on the ground (not sure if this applies to tiles in houses); it will tell you "Both hands have to be free.". If the source container is valid, Tibia will first try to move the item in your right hand (shield) to the source container, then try to move the item in your left hand (weapon) to the source container, then move the two-handed weapon to your left hand.

Equipping a weapon/shield while holding a two-handed weapon is handled the same way. Tibia will only let you equip the item if it's inside a container that you own; otherwise, it will tell you "Drop the double-handed object first.". If the source container and target slot are valid, Tibia will first try to move the two-handed weapon to the source container, then move the weapon/shield to the target slot.

Here is some psuedo-code that should at least solve the moving. Not sure if TFS accounts for the different scenarios I mentioned above:
```
Game::internalMoveItem()
{
    ...
    //check if we can add this item
    ReturnValue ret = toCylinder->queryAdd(index, *item, count, flags, actor);
    if (ret == RETURNVALUE_NEEDEXCHANGE) {
        // The extracted function, that is talked about below, can be applied here since the code used was taken from this block
        ret = extractedFunction(fromCylinder, toCylinder, fromCylinder->getThingIndex(item), toItem);
        if (ret == RETURNVALUE_NOERROR) {
            ret = toCylinder->queryAdd(index, *item, count, flags);
            toItem = nullptr;
        }
    } else if (ret == RETURNVALUE_DROPTWOHANDEDITEM || ret == RETURNVALUE_BOTHHANDSNEEDTOBEFREE) {
        // You may be wondering why I'm checking for both return values when they are two different actions
        // The reason is because they boil down to basically the same action
        // The only difference is that equipping a two-handed weapon will remove the shield and de-equipping a two-handed won't, since there shouldn't be a shield equipped, so it will just skip that step

        // toCylinder should be a Player object, but it's always good to be safe
        // Plus, it makes it easier to work with inventory items
        Player* player = dynamic_cast<Player*>(toCylinder) {
        if (!player) {
            return RETURNVALUE_NOTPOSSIBLE;
        }

        // If there is a right hand (shield) object, we need to move it first
        if (Item* shield = player->getInventoryItem(CONST_SLOT_RIGHT)) {
            // Game::internalMoveItem() already has the code to try to move an item to the source container
            // It should be extracted into a separate function and called here since it needs to be called for the left hand item as well
            // https://github.com/otland/forgottenserver/blob/master/src/game.cpp#L1070
            ret = extractedFunction(fromCylinder, toCylinder, INDEX_WHEREEVER, shield);
            if (ret != RETURNVALUE_NOERROR) {
                return ret;
            }
        }

        // Do the same for the left hand (weapon) object
        if (Item* weapon = player->getInventoryItem(CONST_SLOT_LEFT)) {
            ret = extractedFunction(fromCylinder, toCylinder, INDEX_WHEREEVER, weapon);
            if (ret != RETURNVALUE_NOERROR) {
                return ret;
            }
        }

        // Check if we can now add the item to the destination cylinder
        ret = toCylinder->queryAdd(index, *item, count, flags);
    }
    ...
}
```
Extract function would look something like this. Of course, you'd want to give it a decent name. Also `fromCylinder` and `toCylinder` may be confusing in this sense; better parameter names would be idea.
```
Game::extractedFunction(Cylinder* fromCylinder, Cylinder* toCylinder, int32_t toIndex, Item* item)
{
    ret = fromCylinder->queryAdd(toIndex, *item, item->getItemCount(), 0);
    if (ret == RETURNVALUE_NOERROR) {
        //check how much we can move
        uint32_t maxExchangeQueryCount = 0;
        ReturnValue retExchangeMaxCount = fromCylinder->queryMaxCount(INDEX_WHEREEVER, *item, item->getItemCount(), maxExchangeQueryCount, 0);

        if (retExchangeMaxCount != RETURNVALUE_NOERROR && maxExchangeQueryCount == 0) {
            return retExchangeMaxCount;
        }

        ret = toCylinder->queryRemove(*item, item->getItemCount(), flags);
        if (ret == RETURNVALUE_NOERROR) {
            int32_t oldItemIndex = toCylinder->getThingIndex(item);
            toCylinder->removeThing(item, item->getItemCount());
            fromCylinder->addThing(item);

            if (oldItemIndex != -1) {
                toCylinder->postRemoveNotification(item, fromCylinder, oldItemIndex);
            }

            int32_t newItemIndex = fromCylinder->getThingIndex(item);
            if (newItemIndex != -1) {
                fromCylinder->postAddNotification(item, toCylinder, newItemIndex);
            }
        }
    }
    return ret;
}
```

**Comment by DSpeichert** [comment]
Fixed in efb91ac / #1688
